protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+    bar = 2_000;
+  }
+}
+
+contract SimpleStorage is SimpleStorageLayout {
+  // Whatever runtime logic is needed
+}
+```
+
+With this architecture, the OPCM contract (which will be a non-proxied singleton per release) would
+then either:
+
+1. Deploy a new contract by:
+   1. Deploying a new proxy
+   1. setting its implementation to the `SimpleStoragePopulator`
+   1. calling `initialize()`
+   1. setting its implementation to `SimpleStorage`
+1. Upgrade a contract by:
+   1. setting its implementation to the `SimpleStoragePopulator`
+   1. calling `upgrade()`
+   1. setting its implementation to `SimpleStorage`
+
+This workflow has the benefit of removing the `initializer()` functions, which:
+
+- create a risk of reinit attacks
+- bloat the runtime bytecode size
+- are extremely verbose
+- require finesse to avoid modifying or resetting values which are not meant to be changed.
+
+## A single `SuperchainProxyAdmin`
+
+The existing L1 `ProxyAdmin` (of which there is currently one per OP Chain), will be replaced with a
+single L1 `SuperchainProxyAdmin`. The `SuperchainProxyAdmin`, in contrast with the L1 `ProxyAdmin`,
+will have minimal storage, and will not internally track the type of Proxy which each contract is.
+
+This contract will be owned by the [Upgrade
+Controller](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/stage-1.md#L30), and
+will be a simple pass through contract with a minimal interface.
+
+```
+/// @notice forwards calldata to the specified address.
+///   Reverts unless called by owner.
+function forward(address, bytes) external;
+
+/// @notice forwards calldata to the specified addresses.
+///   Reverts unless called by owner.
+function forward(address[], bytes[]) external;
+```
+
+This architecture makes it possible to manage upgrade authorization for the whole superchain through
+a single storage value in a single contract, rather than having a separate contract per OP Chain.
+
+It also avoids the need for an `sload` to ready the `proxyType` mapping for each contract.
+
+## Identifying OP Chains by the SystemConfig
+
+Rather than tracking the contracts within the `ProxyAdmin`, we will use the `SystemConfig` as the
+unique identifier for each OPChain, and get the contract addresses from it. The custom logic
+required for the two non-standard proxy types will be handled in the OPCM contract, and does not
+require tracking the proxy type, as it is known already for each contract, ie. the only non-standard
+Proxies are the `L1StandardBridge` (`L1ChugsplashProxy`) and `L1CrossDomainMessenger`
+(`ResolvedDelegateProxy`).
+
+## The upgrade process
+
+As part of the release process, the associated implementation contracts and a new OPCM, will be deployed.
+
+The Upgrade Controller MUST be a Safe in order to atomically:
+
+1. transfer ownership of the `SuperchainProxyAdmin` to the new OPCM.
+2. call `OPCM.upgrade()` which will return ownership to the Safe immediately upon completing the
```

Comment:
How is chain specific config passed? Is it calldata? It could also be in storage which feels easier for atomic execution. For example to atomically upgrade all standard chains (some owned by a 2/2 and some owned by a 3/3) we could have a pattern where chains register their config first, then upgrade reads it from storage. For example:

```solidity
contract OPContractsManager {
  mapping (address systemConfig => uint256) params;

  function set(address _systemConfig, uint256 _param) external {
    // This check is not feasble onchain so we'd need a workaround, see https://github.com/ethereum-optimism/client-pod/issues/699
    require(msg.sender == _systemConfig.admin().owner());

    param[_systemConfig] = param;
  }

  function upgrade(address _systemConfig) external { 
    uint256 param = params[_systemConfig];
    // rest of logic here
  }
}
```

Reply:
I had imagined it being in calldata, but I like this pattern a lot for how it enables auditing and clear auth around setting new config values before an `upgrade()`. I'd call it `newParams` just to be clear.

Reply:
Decision: do config params in calldata.
For chains with a different upgrade controller, it's a separate transaction.